auth/proofOfPossession.go

```diff
@@ -0,0 +1,54 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Cloud Security Client Go contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+package auth
+
+// validateCertificate checks proof of possession in addition to audience validation
+// to make sure that it was called by a trust-worthy consumer.
+// Trust between application and applications/services is established with certificates in principle.
+// Proof of possession uses certificates as proof token and therefore, x.509 based mTLS communication is demanded.
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+)
+
+var ErrNoClientCert = errors.New("there is no x509 client certificate provided")
+var ErrNoToken = errors.New("there is no token provided")
+
+// validateCertificate runs all proof of possession checks.
+// This ensures that the token was issued for the sender.
+func validateCertificate(clientCertificate *Certificate, token Token) error {
+	if clientCertificate == nil {
+		return ErrNoClientCert
+	}
+	if token == nil {
+		return ErrNoToken
+	}
+	return ValidateX5tThumbprint(clientCertificate, token)
+}
+
+// ValidateX5tThumbprint compares the thumbprint of the provided X509 client certificate against the cnf claim with the confirmation method "x5t#S256".
+// This ensures that the token was issued for the sender.
+func ValidateX5tThumbprint(clientCertificate *Certificate, token Token) error {
```

Comment:
Should be make it private? As validating the x5t thumbprint should always be followed by the `azp` check according to the concept.

Reply:
But then it can't be called standalone as documented?

Reply:
Yes the question was rather if it actually makes sense for the application to check just the x5t thumbprint. If it does, leave it as is.

Reply:
anyhow, I've set it to private for now, because in case we decide that aud == clientid must be checked as part of the the standalone Certificate Thumbprint (x5t) validator we also would need oauth2config, which is neither provided via parameter nor via context...